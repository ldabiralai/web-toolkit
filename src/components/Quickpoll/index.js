import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'react-emotion';
import * as colors from '../../colors';
import { fontFamilies } from '../../typography';
import { Button } from '../..';

const StyledContainer = styled.div`
  border-radius: 4px;
  background-color: ${colors.flawlessMahogany};
  position: relative;
  width: 100%;
  color: ${colors.coreLightMinus1};
  padding-bottom: 20px;
  padding-top: 10px;
`;

const StyledBackground = styled.div`
  position: absolute;
  height: 100%;
  width: 100%;
  border-radius: 4px;
  background-color: ${colors.coreNeutral9};
  clip-path: polygon(0 11%, 100% 0%, 100% 100%, 0% 100%);
  z-index: 1;
`;

export const StyledTitle = styled.div`
  position: relative;
  font-size: 20px;
  font-family: ${fontFamilies.alphaHeadline};
  margin: 26px 19px;
  z-index: 2;
`;

export const StyledChoices = styled.div`
  position: relative;
  width: 100%;
  z-index: 2;
`;

const StyledChoice = styled(Button)`
  display: flex;
  align-items: center;
  position: relative;
  z-index: 3;
  background-color: ${colors.blackRussian};
  border: 1px solid ${colors.nobel};
  border-radius: 24px;
  margin-bottom: 10px;
  margin-right: 19px;
  margin-left: 19px;
  color: ${colors.coreLightMinus1};
  font-family: ${fontFamilies.interUi};
  font-size: 12px;
  svg {
    display: none;
  }
  ${props =>
    props.showResults &&
    css`
      padding: 2px 0;
      border: 0px;
      border-radius: 0;
      background-color: ${colors.actionOneDarkMinus1};
      height: 48px;
      :hover {
        color: inherit;
        cursor: default;
      }
    `}
`;

const StyledResultBar = styled.div`
  background-color: ${colors.actionOneDarkBase};
  width: ${props => props.result};
  height: 100%;
  position: absolute;
  top: 0;
  z-index: 4;
`;

export const StyledChoiceText = styled.div`
  z-index: 5;
  width: 100%;
`;

export const StyledResultPercentage = styled.div`
  position: relative;
  left: 4%;
  z-index: 5;
`;

class QuickPoll extends React.Component {
  handleClick = id => {
    const { showResults, onChoiceClick } = this.props;
    if (showResults !== true) {
      onChoiceClick(id);
    }
  };

  render() {
    const { title, choices, showResults } = this.props;
    return (
      <StyledContainer>
        <StyledBackground />
        <StyledTitle>{title}</StyledTitle>
        <StyledChoices>
          {choices.map(choice => (
            <StyledChoice
              key={choice.id}
              onClick={() => this.handleClick(choice.id)}
              type="secondary"
              showResults={showResults}
            >
              {showResults && (
                <>
                  <StyledResultPercentage>{choice.result}</StyledResultPercentage>
                  <StyledResultBar result={choice.result} />
                </>
              )}
              <StyledChoiceText>{choice.displayText}</StyledChoiceText>
            </StyledChoice>
          ))}
        </StyledChoices>
      </StyledContainer>
    );
  }
}

QuickPoll.defaultProps = {
  showResults: false,
};

const choiceType = PropTypes.shape({
  id: PropTypes.number,
  displayText: PropTypes.string,
  result: PropTypes.string,
});

QuickPoll.propTypes = {
  title: PropTypes.string.isRequired,
  choices: PropTypes.arrayOf(choiceType).isRequired,
  showResults: PropTypes.bool,
  onChoiceClick: PropTypes.func.isRequired,
};

export default QuickPoll;
