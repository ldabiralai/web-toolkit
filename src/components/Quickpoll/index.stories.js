import React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';
import { text, object, boolean } from '@storybook/addon-knobs';
import { QuickPoll } from '../..';

export const choices = [
  { id: 123451, displayText: 'Nadal' },
  { id: 123452, displayText: 'Djokovic' },
  { id: 123453, displayText: 'Zverev' },
  { id: 123454, displayText: 'Federer' },
  { id: 123455, displayText: 'Other' },
];

export const choicesWithResults = [
  { id: 123451, displayText: 'Nadal', result: '50%' },
  { id: 123452, displayText: 'Djokovic', result: '47%' },
  { id: 123453, displayText: 'Zverev', result: '34%' },
  { id: 123454, displayText: 'Federer', result: '16%' },
  { id: 123455, displayText: 'Other', result: '1%' },
];

const quickPoll = storiesOf('Components|QuickPoll', module).addDecorator(withInfo);

quickPoll.add('QuickPoll', () => (
  <QuickPoll
    title={text('title', 'Who will win the 2019 French Open?')}
    choices={object('choices', choices)}
    // eslint-disable-next-line no-alert
    onChoiceClick={id => alert(`You clicked on ${id}`)}
  />
));

quickPoll.add('QuickPoll with results', () => (
  <QuickPoll
    title={text('title', 'Who will win the 2019 French Open?')}
    choices={object('choices', choicesWithResults)}
    showResults={boolean('showResults', true)}
  />
));
