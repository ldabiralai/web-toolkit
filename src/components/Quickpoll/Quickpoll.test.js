import React from 'react';
import { shallow } from 'enzyme';
import Quickpoll, { StyledChoices, StyledChoiceText } from '.';
import { choices } from './index.stories';

describe('QuickPoll', () => {
  it('renders a QuickPoll Component', () => {
    expect(
      shallow(
        <Quickpoll
          title="Who will win the 2019 French Open?"
          choices={choices}
          showResults={false}
          onChoiceClick={() => null}
        />
      )
    ).toMatchSnapshot();
  });

  it('should display choices', () => {
    const wrapper = shallow(<Quickpoll title="Who will win the 2019 French Open?" choices={choices} />);
    expect(wrapper.find(StyledChoices).children()).toHaveLength(5);
    expect(
      wrapper
        .find(StyledChoiceText)
        .first()
        .debug()
    ).toEqual('Nadal');
  });
});
