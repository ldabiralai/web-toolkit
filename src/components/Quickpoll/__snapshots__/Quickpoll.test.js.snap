// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuickPoll renders a QuickPoll Component 1`] = `
<Styled(div)>
  <Styled(div) />
  <Styled(div)>
    Who will win the 2019 French Open?
  </Styled(div)>
  <Styled(div)>
    <Styled(ArrowLink)
      key="123451"
      onClick={[Function]}
      showResults={false}
      type="secondary"
    >
      <Styled(div)>
        Nadal
      </Styled(div)>
    </Styled(ArrowLink)>
    <Styled(ArrowLink)
      key="123452"
      onClick={[Function]}
      showResults={false}
      type="secondary"
    >
      <Styled(div)>
        Djokovic
      </Styled(div)>
    </Styled(ArrowLink)>
    <Styled(ArrowLink)
      key="123453"
      onClick={[Function]}
      showResults={false}
      type="secondary"
    >
      <Styled(div)>
        Zverev
      </Styled(div)>
    </Styled(ArrowLink)>
    <Styled(ArrowLink)
      key="123454"
      onClick={[Function]}
      showResults={false}
      type="secondary"
    >
      <Styled(div)>
        Federer
      </Styled(div)>
    </Styled(ArrowLink)>
    <Styled(ArrowLink)
      key="123455"
      onClick={[Function]}
      showResults={false}
      type="secondary"
    >
      <Styled(div)>
        Other
      </Styled(div)>
    </Styled(ArrowLink)>
  </Styled(div)>
</Styled(div)>
`;
